meeko/molsetup.py

```diff
@@ -4,623 +4,451 @@
-            nr_params_to_add = nr_params_to_add.pop()
-            all_ok &= ok
-            all_err += err
-            for key in p:
-                atom_params.setdefault(key, [None]*atom_counter)
-                atom_params[key].extend(p[key])
-            atom_counter += nr_params_to_add
-            for key in atom_params:
-                if key not in p:
-                    atom_params[key].extend([None]*nr_params_to_add)
-    
-        if not all_ok:
-            raise RuntimeError(all_err)
-        
-        molsetup = cls()
-        charges = atom_params.pop("gasteiger")
-        atypes = atom_params.pop("atom_types")
-        for i in range(len(x)):
-            molsetup.add_atom(
-                i,
-                coord = (x[i], y[i], z[i]),
-                charge = charges[i],
-                atom_type = atypes[i],
-                element = None,
-                pdbinfo = pdbinfo,
-                chiral = False,
-                ignore = False,
-            )
+        # If atom index is not specified, appends the new atom to the end of the current atom list
+        if atom_index is None:
+            atom_index = len(self.atoms)
+
+        # Inserts dummy atoms if a specified atom index is greater than the current length of the atom list
+        while atom_index > len(self.atoms):
+            self.atoms.append(Atom(len(self.atoms), is_dummy=True))
+
+        # Creates and adds new atom to the atom list
+        new_atom = Atom(
+            atom_index, pdbinfo, charge, element, atom_type, is_ignore, is_chiral, graph
+        )
+        if atom_index < len(self.atoms):
+            self.atoms[atom_index] = new_atom
+            return
+        self.atoms.append(new_atom)
+        return
 
-        molsetup.atom_params = atom_params
+    def add_pseudo_atom(
+        self,
+        pdbinfo: str = DEFAULT_PDBINFO,
+        charge: float = DEFAULT_CHARGE,
+        atom_type: str = DEFAULT_ATOM_TYPE,
+        is_ignore: bool = DEFAULT_IS_IGNORE,
+        anchor_list: list[int] = None,
+        rotatable: bool = False,
+        directional_vectors: list[int] = None,
+    ):
+        """
 
-        return molsetup 
+        Parameters
+        ----------
+        pdbinfo
+        charge
+        atom_type
+        is_ignore
+        anchor_list
+        rotatable
+        directional_vectors
 
-    def set_types_from_uniq_atom_params(self, uniq_atom_params, prefix):
-        param_idxs = uniq_atom_params.get_indices_from_atom_params(self.atom_params)
-        if len(param_idxs) != len(self.atom_type):
-            raise RuntimeError("{len(param_idxs)} parameters but {len(self.atom_type)} in molsetup")
-        for i, j in enumerate(param_idxs):
-            self.atom_type[i] = f"{prefix}{j}"
-        return None
+        Returns
+        -------
 
+        """
+        # Places the atom at the end of the atom list.
+        pseudoatom_index = len(self.atoms)
```

Comment:
Earlier, the atom index assigned to a pseudoatom was the true atom count + the number of pseudoatoms, because of how the dictionary structure for the attributes worked. I changed it to this because this seemed to be the intention, is there any reason to change it back/to preserve fields like true_atom_count if we're not doing this?

Reply:
Having the `is_pseudo` flag suffices. The count of "true" atoms is then the number of atoms which are not pseudos. I also think we need only one of `is_pseudo` and `is_dummy` as these are synonyms.